plugins/storage/server/api/files.ts

```diff
@@ -0,0 +1,100 @@
+import {
+  closeSync,
+  createReadStream,
+  createWriteStream,
+  existsSync,
+  mkdirSync,
+  openSync,
+} from "fs";
+import path from "path";
+import Router from "koa-router";
+import mime from "mime-types";
+import { bytesToHumanReadable } from "@shared/utils/files";
+import env from "@server/env";
+import {
+  AuthenticationError,
+  InvalidRequestError,
+  NotFoundError,
+} from "@server/errors";
+import auth from "@server/middlewares/authentication";
+import { Attachment } from "@server/models";
+import { authorize } from "@server/policies";
+import { APIContext } from "@server/types";
+import { getAttachmentForJWT } from "@server/utils/jwt";
+import { getFileFromRequest } from "@server/utils/koa";
+import { assertPresent, validateKey } from "@server/validation";
+
+const router = new Router();
+
+router.post("files.create", auth(), async (ctx: APIContext) => {
+  if (!ctx.is("multipart/form-data")) {
+    throw InvalidRequestError("Request type must be multipart/form-data");
+  }
+  const { key } = ctx.request.body;
+  validateKey(key);
+  const file = getFileFromRequest(ctx.request);
+  if (!file) {
+    throw InvalidRequestError("Request must include a file parameter");
+  }
+  if (
+    env.FILE_STORAGE_UPLOAD_MAX_SIZE &&
+    file.size > env.FILE_STORAGE_UPLOAD_MAX_SIZE
+  ) {
+    throw InvalidRequestError(
+      `The selected file was larger than the ${bytesToHumanReadable(
+        env.FILE_STORAGE_UPLOAD_MAX_SIZE
+      )} maximum size`
+    );
+  }
+
+  const subdir = key.split("/").slice(0, -1).join("/");
```

Comment:
All this kind of stuff should be inside `LocalStorage` rather than the router.

Reply:
What if we expose this through a method on `Attachment`, something like `attachment.saveFile(file)` which internally calls the `FileStorage.upload`? Makes it consistent with other getters like `attachment.stream` and `attachment.buffer`.

Reply:
sounds good